highfidelity/hifi
# Adding support for variable allocated textures to the GL 4.1 backend

This PR adds the same dynamic re-allocation functionality currently used for the GL 4.5 backend to the GL 4.1 backend.  A significant amount of the code for managing the variable allocated textures has been moved to a common support class.  

## Testing

Testing should be done using the [Variable Allocated Textures](https://highfidelity.atlassian.net/wiki/display/HOME/Variable+Allocated+Textures) test plan in conjunction with the [instructions](https://highfidelity.atlassian.net/wiki/display/HOME/Testing+OpenGL+4.1+functionality) for forcing interface to use the OpenGL 4.1 backend.  

The application should also be tested in the same way without forcing the OpenGL 4.1 backend, as some code for the OpenGL 4.5 implementation has moved and is now shared by both implementations.  

## OpenGL 4.1 testing

* Set an environment variable on your computer with the name HIFI_DISABLE_OPENGL_45.  The value of the environment variable does not matter, so long as it's set.  Once you've set this variable, you can start Interface
* Open Interface
* Verify that you are using the 4.1 back-end by opening up the log file viewer (Ctrl-Shift-L) and entering the text Using OpenGL in the search field.  If you have successfully set the environment variable then you should see a line that says `Using OpenGL 4.1 backend`
* Perform the testing for variable allocated textures 
* Make sure when you are finished testing to remove the HIFI_DISABLE_OPENGL_45 environment variable, or it could negatively impact your performance on a system that would otherwise be using the OpenGL 4.5 backend

## Variable Allocated Textures Testing

For testing it is desirable to have the stats display (which can toggled with the '/' key) turned on, as it includes a real time display of the currently consumed texture memory (listed as `Resource Memory`) as well as the amount of texture information which is waiting to be populated (listed as `Pending Transfer`).  When testing it's best to wait until the current state is stable and the Pending Transfer value is zero before making observations.  
To test the variable allocated texture functionality...

### Setup

* Start Interface
* Ensure that the current allowed memory is set to it's default value by going into `Developer | Render | Maximum Texture Memory` menu and ensuring that Automatic Texture Memory is selected
* Go to a domain with lots of content that includes many textures.  Typically `Playa or `Dev-Playa` are ideal as they consume a significant amount of texture memory.
* Spin the viewpoint around a couple of times to ensure that most of the domain content is loaded into memory
* View the stats display and wait for the `Pending Transfer` value to fall to zero.  The `Resource Memory` value should be at least 512 MB.
* Position yourself near one of the models in the scene with a high quality texture, such as directly in front of the Circus Casino sign

### Over-subscription testing

* Open the `Developer | Render | Maximum Texture Memory` menu and select 64 MB
* The `Resource Memory` value should rapidly fall to less than 64 MB.  Concurrently the quality of the textures in the rendered scene should visibly drop

### Under-subscription testing

* Open the `Developer | Render | Maximum Texture Memory` menu and reset it to Automatic Texture Memory
* The `Resource Memory` value should rapidly rise to it's previous value, and the `Pending Transfer` value should jump up to nearly that amount as well.
* The `Pending Transfer` value should then start to fall.  As it does so, the textured items in the scene should return to their previous quality.
